File: src/adapters/git/GitLabAdapter.ts
import axios from 'axios';

import { GitAdapter } from './GitAdapter';

export interface GitLabMRParams {
  projectId: string;
  mergeRequestIid: string;
}

/**
 * Represents a parsed GitLab URL
 */
interface ParsedGitLabUrl {
  origin: string;
  projectId: string;
  mergeRequestIid: string;
}

/**
 * Represents a GitLab file change
 */
interface GitLabFileChange {
  new_path: string;
  diff: string;
}

/**
 * Adapter for GitLab platform
 */
export class GitLabAdapter extends GitAdapter {
  /**
   * Parse a GitLab MR URL to extract project ID and MR IID
   * @param url - The URL to parse
   * @returns Object containing projectPath and mergeRequestIid
   * @throws If URL is invalid or cannot be parsed
   */
  parseRequestUrl(url: string): Record<string, string> {
    try {
      const parsedUrl = this.parseUrl(url);

      // Set the gitlabUrl in the config to the origin from the URL
      // This ensures we're using the correct GitLab instance
      (this as any).config.gitlabUrl = parsedUrl.origin;

      return {
        projectId: parsedUrl.projectId,
        mergeRequestIid: parsedUrl.mergeRequestIid,
      };
    } catch (error) {
      console.error('Error parsing GitLab MR URL:', (error as Error).message);
      throw new Error(`Invalid GitLab MR URL: ${url}`);
    }
  }

  /**
   * Parse a GitLab URL into its components
   *
   * @param url URL to parse
   * @returns Parsed URL components
   */
  private parseUrl(url: string): ParsedGitLabUrl {
    const parsedUrl = new URL(url);
    const pathParts = parsedUrl.pathname.split('/');

    const mergeRequestIid = this.extractMergeRequestId(pathParts);
    const projectId = this.extractProjectId(pathParts);

    return {
      origin: parsedUrl.origin,
      projectId,
      mergeRequestIid,
    };
  }

  /**
   * Extract the merge request ID from URL path parts
   *
   * @param pathParts Split path components
   * @returns Merge request ID
   */
  private extractMergeRequestId(pathParts: string[]): string {
    const mrIndex = pathParts.findIndex(part => part === 'merge_requests');

    if (mrIndex === -1 || mrIndex + 1 >= pathParts.length) {
      throw new Error('Invalid GitLab MR URL format');
    }

    const mergeRequestId = pathParts[mrIndex + 1];
    if (mergeRequestId === undefined) {
      throw new Error('Merge request ID not found in URL');
    }

    return mergeRequestId;
  }

  /**
   * Extract project ID from URL path parts
   *
   * @param pathParts Split path components
   * @returns Project ID or encoded project path
   */
  private extractProjectId(pathParts: string[]): string {
    const mrIndex = pathParts.findIndex(part => part === 'merge_requests');
    if (mrIndex === -1) {
      throw new Error('Could not find merge_requests in URL');
    }

    // Find the position of '-' before 'merge_requests'
    const dashIndex = this.findDashIndex(pathParts, mrIndex);

    // Skip the first empty element if pathname starts with /
    const startIndex = pathParts[0] === '' ? 1 : 0;

    // Get project path based on whether we found a dash
    const projectPath = this.buildProjectPath(pathParts, startIndex, dashIndex, mrIndex);

    return this.normalizeProjectId(projectPath);
  }

  /**
   * Find the dash index before merge_requests
   *
   * @param pathParts Split path components
   * @param mrIndex Index of merge_requests
   * @returns Index of dash or -1 if not found
   */
  private findDashIndex(pathParts: string[], mrIndex: number): number {
    for (let i = 0; i < mrIndex; i++) {
      if (pathParts[i] === '-') {
        return i;
      }
    }
    return -1;
  }

  /**
   * Build the project path from path parts
   *
   * @param pathParts Split path components
   * @param startIndex Starting index
   * @param dashIndex Index of dash
   * @param mrIndex Index of merge_requests
   * @returns Project path
   */
  private buildProjectPath(
    pathParts: string[],
    startIndex: number,
    dashIndex: number,
    mrIndex: number
  ): string {
    if (dashIndex > 0) {
      // If we found a dash, extract the path up to the dash
      return pathParts.slice(startIndex, dashIndex).join('/');
    }
    // If no dash, take everything before 'merge_requests'
    return pathParts.slice(startIndex, mrIndex - 1).join('/');
  }

  /**
   * Normalize project ID by either using directly if numeric or encoding
   *
   * @param projectPath Raw project path from URL
   * @returns Normalized project ID
   */
  private normalizeProjectId(projectPath: string): string {
    // Try to determine if the project path is numeric (a project ID) or a path
    if (/^\d+$/.test(projectPath)) {
      // If it's just a number, use it directly as a project ID
      return projectPath;
    }
    // Otherwise URL encode the path for the API
    return encodeURIComponent(projectPath || '');
  }

  /**
   * Get the diff of a merge request from GitLab
   * @param params - Parameters with projectId and mergeRequestIid
   * @returns The diff content
   */
  async getRequestDiff(params: Record<string, string>): Promise<string> {
    const glParams = params as unknown as GitLabMRParams;
    try {
      const response = await this.fetchMergeRequestChanges(glParams);
      return this.formatDiffFromChanges(response.data.changes);
    } catch (error) {
      return this.handleDiffError(error);
    }
  }

  /**
   * Fetch merge request changes from GitLab API
   *
   * @param params GitLab MR parameters
   * @returns API response
   */
  private async fetchMergeRequestChanges(params: GitLabMRParams) {
    const url = this.buildMergeRequestUrl(params, 'changes');
    return await axios.get(url, {
      headers: this.getRequestHeaders(),
    });
  }

  /**
   * Build GitLab API URL for merge request
   *
   * @param params GitLab MR parameters
   * @param endpoint API endpoint
   * @returns Complete API URL
   */
  private buildMergeRequestUrl(params: GitLabMRParams, endpoint: string): string {
    return `${(this as any).config.gitlabUrl}/api/v4/projects/${params.projectId}/merge_requests/${params.mergeRequestIid}/${endpoint}`;
  }

  /**
   * Get request headers for GitLab API
   *
   * @returns Headers object with token
   */
  private getRequestHeaders() {
    return {
      'PRIVATE-TOKEN': (this as any).config.gitlabToken || '',
    };
  }

  /**
   * Format diff text from GitLab file changes
   *
   * @param changes Array of file changes
   * @returns Formatted diff text
   */
  private formatDiffFromChanges(changes: GitLabFileChange[]): string {
    return changes.map(change => `File: ${change.new_path}\n${change.diff}`).join('\n\n');
  }

  /**
   * Handle errors from diff request
   *
   * @param error Error from API request
   * @throws Original error after logging
   */
  private handleDiffError(error: unknown): never {
    console.error('Error fetching MR diff from GitLab:', (error as Error).message);

    if (axios.isAxiosError(error) && error.response) {
      console.error('Response status:', error.response.status);
      console.error('Response data:', error.response.data);
    }

    throw error;
  }

  /**
   * Post a comment on a GitLab merge request
   * @param params - Parameters with projectId and mergeRequestIid
   * @param comment - The comment content
   */
  async commentOnRequest(params: Record<string, string>, comment: string): Promise<void> {
    const glParams = params as unknown as GitLabMRParams;
    try {
      await this.postMergeRequestComment(glParams, comment);
    } catch (error) {
      this.handleCommentError(error);
    }
  }

  /**
   * Post comment to GitLab API
   *
   * @param params GitLab MR parameters
   * @param comment Comment text
   */
  private async postMergeRequestComment(params: GitLabMRParams, comment: string): Promise<void> {
    const url = this.buildMergeRequestUrl(params, 'notes');
    const formattedComment = this.formatComment(comment);

    await axios.post(url, { body: formattedComment }, { headers: this.getRequestHeaders() });
  }

  /**
   * Format comment text with heading
   *
   * @param comment Raw comment text
   * @returns Formatted comment text
   */
  private formatComment(comment: string): string {
    return `## LLM Code Review Feedback\n\n${comment}`;
  }

  /**
   * Handle errors from comment request
   *
   * @param error Error from API request
   * @throws Original error after logging
   */
  private handleCommentError(error: unknown): never {
    console.error('Error posting comment on MR:', (error as Error).message);

    if (axios.isAxiosError(error) && error.response) {
      console.error('Response status:', error.response.status);
      console.error('Response data:', error.response.data);
    }

    throw error;
  }
}

File: python/tiw/adapters/git/gitlab_adapter.py
"""GitLab API adapter for MR operations."""

import re
from typing import Dict, Optional, Union

import gitlab

from ...config.config import AppConfig
from .git_adapter import GitAdapter


class GitLabAdapter(GitAdapter):
    """Adapter for GitLab API."""
    
    def __init__(self, config: AppConfig):
        """Initialize the GitLab adapter.
        
        Args:
            config: Application configuration
        """
        super().__init__(config)
        
        # Initialize the client
        self.gitlab = self._init_client()
    
    def _init_client(self) -> gitlab.Gitlab:
        """Initialize the GitLab API client.
        
        Returns:
            GitLab client
            
        Raises:
            ValueError: If token is missing
        """
        if not self.config.gitlab_token:
            raise ValueError("GitLab token is required")
        
        return gitlab.Gitlab(self.config.gitlab_url, private_token=self.config.gitlab_token)
    
    def parse_request_url(self, url: str) -> Dict[str, str]:
        """Parse a GitLab MR URL into project ID and MR IID.
        
        Args:
            url: The GitLab MR URL
            
        Returns:
            Dictionary with projectId and mergeRequestIid
            
        Raises:
            ValueError: If URL is invalid
        """
        # GitLab MR URL format: https://gitlab.com/group/project/-/merge_requests/123
        pattern = r'https?://[^/]+/(.+?)/-/merge_requests/(\d+)'
        match = re.match(pattern, url)
        
        if not match:
            # Try alternative format without '-/'
            alt_pattern = r'https?://[^/]+/(.+?)/merge_requests/(\d+)'
            match = re.match(alt_pattern, url)
            
            if not match:
                raise ValueError(f"Invalid GitLab MR URL: {url}")
        
        return {
            "projectId": match.group(1),
            "mergeRequestIid": match.group(2),
        }
    
    async def get_request_diff(self, params: Dict[str, str]) -> str:
        """Get the diff of a GitLab MR.
        
        Args:
            params: Parameters needed to identify the MR (projectId, mergeRequestIid)
            
        Returns:
            The diff content
            
        Raises:
            ValueError: If required parameters are missing
            Exception: If GitLab API request fails
        """
        # Validate required params
        if not all(key in params for key in ["projectId", "mergeRequestIid"]):
            raise ValueError("Missing required parameters: projectId, mergeRequestIid")
        
        try:
            # Get project and merge request
            project = self.gitlab.projects.get(params["projectId"])
            mr = project.mergerequests.get(int(params["mergeRequestIid"]))
            
            # Get the changes
            changes = mr.changes()
            
            # Format as a unified diff
            diff_content = []
            for change in changes.get("changes", []):
                filename = change.get("new_path", change.get("old_path", ""))
                
                if self.config.ignore_lock_files and self._is_lock_file(filename):
                    continue
                
                # Add the diff for this file
                if change.get("diff"):
                    diff_content.append(change["diff"])
            
            return "\n".join(diff_content)
        
        except gitlab.exceptions.GitlabError as e:
            raise Exception(f"GitLab API error: {str(e)}")
        except Exception as e:
            raise Exception(f"Error getting MR diff: {str(e)}")
    
    def _is_lock_file(self, filename: str) -> bool:
        """Check if a file is a lock file.
        
        Args:
            filename: Filename to check
            
        Returns:
            True if it's a lock file
        """
        return filename.endswith(('.lock', 'lock.json', 'package-lock.json', 'yarn.lock', 'pnpm-lock.yaml'))
    
    async def comment_on_request(self, params: Dict[str, str], comment: str) -> None:
        """Post a comment on a GitLab MR.
        
        Args:
            params: Parameters needed to identify the MR (projectId, mergeRequestIid)
            comment: The comment content
            
        Raises:
            ValueError: If required parameters are missing
            Exception: If GitLab API request fails
        """
        # Validate required params
        if not all(key in params for key in ["projectId", "mergeRequestIid"]):
            raise ValueError("Missing required parameters: projectId, mergeRequestIid")
        
        try:
            # Get project and merge request
            project = self.gitlab.projects.get(params["projectId"])
            mr = project.mergerequests.get(int(params["mergeRequestIid"]))
            
            # Post the comment
            mr.notes.create({"body": comment})
        
        except gitlab.exceptions.GitlabError as e:
            raise Exception(f"GitLab API error: {str(e)}")
        except Exception as e:
            raise Exception(f"Error posting comment: {str(e)}")